README.md
# mcrouter
[![Build Status](https://travis-ci.org/evertrue/mcrouter-cookbook.svg)](https://travis-ci.org/evertrue/mcrouter-cookbook)

Install [mcrouter](https://github.com/facebook/mcrouter) and its dependencies, and provide mechanisms to configure and start mcrouter.

This cookbook can, optionally, install a copy of memcached to use locally. This can be disabled by setting `node['mcrouter']['local_memcached']` to false.

## Requirements

### Cookbooks

* `memcached` cookbook
* `magic` cookbook

### Platforms

The following platforms are supported and tested with Test Kitchen:

* Ubuntu 16.04
* Ubuntu 18.04

### Chef

* Chef 12.11+

## Usage

* Include `mcrouter::default` in your node’s run list.

If you wish to customize any of the configuration for mcrouter, you will want to edit the two attributes used to populate the config file and the CLI arguments passed to `mcrouter` on service start:

* ``node['mcrouter']['cli_opts']``
    - Do not overwrite this completely; the values specified already are required, instead, either:
        + supply updated values for these in addition to your new options
        + specify your particular CLI option: `default['mcrouter']['cli_opts']['async-dir']`
* ``node['mcrouter']['config']`` - See configuration options at [mcrouter wiki](https://github.com/facebook/mcrouter/wiki).

## Attributes

* `node['mcrouter']['user']` - User that mcrouter will run as

* `node['mcrouter']['local_memcached']` - Enable a local installation of memcache. `true` by default.
* `node['memcached']['port']` - Configure memcached port. This are overrides of [memcached cookbook](https://github.com/chef-cookbooks/memcached)
* `node['memcached']['udp_port']` - Configure memcached udp port. This are overrides of [memcached cookbook](https://github.com/chef-cookbooks/memcached)

* `node['mcrouter']['cli_opts']` - Collection of hashes that will be passed to `mcrouter` on start. You can see the whole list [here](https://github.com/facebook/mcrouter/wiki/Command-line-options)
* `node['mcrouter']['config']` - This will become mcrouter configuration file. You can read more at [Config Files on mcrouter wiki](https://github.com/facebook/mcrouter/wiki/Config-Files)

## Authors

Author:: EverTrue, Inc. <devops@evertrue.com>  
Author:: Jeff Byrnes <thejeffbyrnes@gmail.com>


CHANGELOG.md
# CHANGELOG for mcrouter

## [v3.0.1] (2019-03-26)

* Fix CHANGELOG to accurately reflect v3.0.0

## [v3.0.0] (2019-03-26)

### Breaking changes:

* Drop source-based installation in favor of APT package installation
* Drop support for Ubuntu 14.04
* Switch to SystemD for the mcrouter service

### Changes:

* Add support for Ubuntu 16.04 & Ubuntu 18.04

## [v2.0.2] (2015-12-15)

### Fixes:

* Delete only build subdirectory (so that ark does not run twice)

## [v2.0.1] (2015-12-07)

### Fixes:

* Updated test kitchen file format
* Re-order build/install ops so that build dirs are always deleted

## [v2.0.0] (2015-08-17)

### Breaking changes:

* Folly version updated to v0.53.0
    - This is now adjustable via an attribute, `node['folly']['version']`
* Mcrouter version set to v0.5.0
    - This is now adjustable via an attribute, `node['mcrouter']['version']`

### Fixes

* Packages needed for Folly/Mcrouter updated as per most recent Folly/Mcrouter docs

## [v1.0.1] (2015-07-15)

### Fixes:

* Add dependency package: libssl-dev

## [v1.0.0] (2015-07-09)

### Breaking changes:

* `['mcrouter']['src_dir']` no longer determines build location (that is now `:file_cache_path`), and as a result...
* Builds are now done in `:file_cache_path` instead of `/opt`
* `mcrouter::folly` is now included by `mcrouter::install`

### Changes:

* Vastly clean up the install process (e.g. now using `ark` for source install)
    - Stopped installing gtest
    - Pruned Apt package install list
    - Use `./configure` defaults rather than specifying a zillion environment variables
    - Just do `make install` for mcrouter and use the default install location (rather than linking `/usr/local/bin/mcrouter -> node['mcrouter']['install_dir']/bin/mcrouter`)
    - Just install `build-essentials` with the cookbook rather than devoting half the cookbook to setting up the build system
    - Break package installs out into `_deps` recipe
* Folly and McRouter now delete their build directories once they're finished installing
* Update the README
* Clean up the Upstart script a bit

### Fixes:

* Run `ldconfig` after `make install` on folly (addresses "library not found" issue with mcrouter)

## [v0.2.2] (2015-02-28)

### Fixes:

* Subscribed `service[mcrouter]` to its own service script to restart on any changes to its CLI options

## [v0.2.1] (2015-02-25)

### Fixes:

* Updated, clarified README

## [v0.2.0] (2015-02-25)

### Fixes:

* Replace missed hard-coded ownership settings with `node['mcrouter']['user']`

### Changes:

* Switch from hard-coded CLI flags to an hash attribute-driven system, utilizing the [`magic` cookbook](https://github.com/sczizzo/magic-cookbook)

## [v0.1.0] (2015-02-24)

### Fixes:

* Fix paths for dependent libraries for `mcrouter` build
* Ensure dirs needed for `mcrouter` exist
* Fix test assertion for `mcrouter` config

### Changes:

* Parameterize paths for folly & mcrouter
    - Makes for simpler configuration of builds
* Fix path to optionally-installed double-conversion lib
* Install & configure `memcached` to start on port 11811
    - `mcrouter` needs an instance of `memcached` somewhere to talk to; simplest thing is to just install it locally
* Set up an Upstart service to manage `mcrouter`

## [v0.0.1] (2015-02-19)

### Changes:

* Initial release:
    - Builds & installs mcrouter

[v3.0.1]: https://github.com/evertrue/mcrouter-cookbook/compare/v3.0.0...v3.0.1
[v3.0.0]: https://github.com/evertrue/mcrouter-cookbook/compare/v2.0.2...v3.0.0
[v2.0.1]: https://github.com/evertrue/mcrouter-cookbook/compare/v2.0.0...v2.0.1
[v2.0.0]: https://github.com/evertrue/mcrouter-cookbook/compare/v1.0.1...v2.0.0
[v1.0.1]: https://github.com/evertrue/mcrouter-cookbook/compare/v1.0.0...v1.0.1
[v1.0.0]: https://github.com/evertrue/mcrouter-cookbook/compare/v0.2.2...v1.0.0
[v0.2.2]: https://github.com/evertrue/mcrouter-cookbook/compare/v0.2.1...v0.2.2
[v0.2.1]: https://github.com/evertrue/mcrouter-cookbook/compare/v0.2.0...v0.2.1
[v0.2.0]: https://github.com/evertrue/mcrouter-cookbook/compare/v0.1.0...v0.2.0
[v0.1.0]: https://github.com/evertrue/mcrouter-cookbook/compare/v0.0.1...v0.1.0
[v0.0.1]: https://github.com/evertrue/mcrouter-cookbook/compare/da547ce9...v0.0.1


recipes/default.rb
#
# Cookbook Name:: mcrouter
# Recipe:: default
#
# Copyright 2015 EverTrue, Inc.
# Copyright 2019 Jeff Byrnes
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

unless platform?('ubuntu') && ::Gem::Version.new(node['platform_version']) >= ::Gem::Version.new(16.04)
  Chef::Log.fatal 'This cookbook only supports Ubuntu 16.04 & 18.04'
end

apt_repository 'mcrouter' do
  uri "https://facebook.github.io/mcrouter/debrepo/#{node['lsb']['codename']}"
  components %w(contrib)
  deb_src false
  key "https://facebook.github.io/mcrouter/debrepo/#{node['lsb']['codename']}/PUBLIC.KEY"
end

apt_update
build_essential 'mcrouter'

include_recipe 'memcached::default' if node['mcrouter']['local_memcached']

user node['mcrouter']['user'] do
  system true
  shell '/bin/false'
end

package 'mcrouter'

cli_opts = node['mcrouter']['cli_opts']

[
  '/etc/mcrouter',
  File.dirname(cli_opts['log-path']),
  cli_opts['async-dir'],
  cli_opts['stats-root'],
].each do |dir|
  directory dir do
    owner node['mcrouter']['user']
    group node['mcrouter']['user']
    recursive true
  end
end

file cli_opts['config-file'] do
  content json_config(node['mcrouter']['config'])
  owner node['mcrouter']['user']
  group node['mcrouter']['user']
end

systemd_unit 'mcrouter.service' do
  content(
    Unit: {
      Description: 'mcrouter',
      After: 'network.target',
    },
    Service: {
      User: node['mcrouter']['user'],
      ExecStart: "#{node['mcrouter']['bin_path']} #{shell_opts(node['mcrouter']['cli_opts'])}",
      Restart: 'on-failure',
    },
    Install: {
      WantedBy: 'multi-user.target',
    }
  )
  action :create
end

service 'mcrouter' do
  provider Chef::Provider::Service::Systemd
  action %i(enable restart)
end
